strawberry/http/async_base_view.py

```diff
@@ -201,15 +220,99 @@ async def run(
             response_data=response_data, sub_response=sub_response
         )
 
+    def encode_multipart_data(self, data: Any, separator: str) -> str:
+        return "".join(
+            [
+                f"\r\n--{separator}\r\n",
+                "Content-Type: application/json\r\n\r\n",
+                self.encode_json(data),
+                "\n",
+            ]
+        )
+
+    def _stream_with_heartbeat(
+        self, stream: Callable[[], AsyncGenerator[str, None]]
+    ) -> Callable[[], AsyncGenerator[str, None]]:
+        """Adds a heartbeat to the stream, to prevent the connection from closing
+        when there are no messages being sent."""
+        # TODO: handle errors
+        # TODO: should we do this more efficiently? and only send the heartbeat when
+        # 5 seconds have passed after the last message? (apollo router doesn't seem to do this)
+        queue = asyncio.Queue[Tuple[bool, Any]](1)
+
+        cancelling = False
+
+        async def drain():
+            try:
+                async for item in stream():
+                    await queue.put((False, item))
+            except Exception as e:
+                if not cancelling:
+                    await queue.put((True, e))
+                else:
+                    raise
+
+        async def heartbeat():
+            while True:
+                await queue.put((False, self.encode_multipart_data({}, "graphql")))
+
+                await asyncio.sleep(5)
+
+        async def merged() -> AsyncGenerator[str, None]:
+            heartbeat_task = asyncio.create_task(heartbeat())
+            task = asyncio.create_task(drain())
+
+            def cancel_tasks():
+                nonlocal cancelling
+                cancelling = True
+                task.cancel()
+                heartbeat_task.cancel()
```

Comment:
If I remember correctly, those needs to be awaited after cancelling:

```suggestion
                task.cancel()
                with contextlib.suppress(asyncio.CancelledError):
                    await task

                heartbeat_task.cancel()
                with contextlib.suppress(asyncio.CancelledError):
                    await heartbeat_task
```